lib/swaps/Swaps.ts

```diff
@@ -243,16 +243,21 @@ class Swaps extends EventEmitter {
    * A promise wrapper for a swap procedure
    * @param maker the remote maker order we are filling
    * @param taker our local taker order
-   * @returns A promise that is resolved once the swap is completed, or rejects otherwise
+   * @returns A promise that resolves to a [[SwapSuccess]] once the swap is completed, or a [[SwapFailureReason]] if it fails
    */
-  public executeSwap = async (maker: PeerOrder, taker: OwnOrder): Promise<SwapSuccess> => {
+  public executeSwap = async (maker: PeerOrder, taker: OwnOrder): Promise<SwapSuccess | SwapFailureReason> => {
     await this.verifyExecution(maker, taker);
-    const rHash = await this.beginSwap(maker, taker);
-    if (!rHash) {
-      throw new Error('cannot execute swap. rHash not found');
+    const beginResult = await this.beginSwap(maker, taker);
+    let rHash: string;
+    if (typeof beginResult === 'string') {
+      // we were able to request the swap and got its rHash value
+      rHash = beginResult;
+    } else {
+      // we could not begin the swap, return the failure reason
+      return beginResult;
     }
 
-    return new Promise<SwapSuccess>((resolve, reject) => {
+    return new Promise<SwapSuccess | SwapFailureReason>((resolve) => {
```

Comment:
Why not use the failure case as a promise reject, so you wont need to deal with the Success/Failure value inspection?

Reply:
I thought about this, but I was unsure about relying on the parameter type in the `catch` blocks we'd need to handle rejections. I'm pretty sure that by default the compiler always treats the error in a `catch` block as `any`. I could explicitly define it as a `SwapFailureReason` though. I was just concerned about what would happen should any other type of exception get thrown for whatever reason.

If you think that's not a big problem though, I can definitely restructure these to throw/reject `SwapFailureReason` - this could also solve the need for anything like a `isSwapBeginSuccess` above because `beginSwap` could also throw the failure reason.

Reply:
Now that I understand the limitation, i'd say you should go with whatever you think is better.

You can perhaps move the inspection to the thrown value, so the method interface will be straightforward. Perhaps by defining and throwing `SwapFailure`, an object with one property, `reason`, unlike `SwapFailureReason` which is a plain number.